### PR TITLE
Add section to coding guidelines on Optional usage

### PR DESCRIPTION
**Pull request by trask**
This was discussed a long while back, but never documented (and recently came up in #6889)

### CODE REVIEW

**Review by mateuszrzeszutek — APPROVED** [review]

File: docs/contributing/style-guideline.md [hunk 1]
```diff
@@ -126,3 +126,11 @@ plugins {
...
 }
 ```
+
+## java.util.Optional usage
+
+`Optional` is not generally used in this project, following the reasoning from
+https://speakerdeck.com/trustin/writing-a-java-library-with-better-experience?slide=12.
+
+It is ok to use `Optional` in places where it does not leak into public API signatures and where
+performance is not critical.
```
Comment (theletterf) [inline docs/contributing/style-guideline.md:136]:
```suggestion
Use `Optional` where it does not leak into public API signatures and where
performance is not a critical aspect.
```

File: docs/contributing/style-guideline.md [hunk 2]
```diff
@@ -126,3 +126,11 @@ plugins {
   id("otel.nullaway-conventions")
 }
 ```
+
+## java.util.Optional usage
+
+`Optional` is not generally used in this project, following the reasoning from
+https://speakerdeck.com/trustin/writing-a-java-library-with-better-experience?slide=12.
```
Comment (theletterf) [inline docs/contributing/style-guideline.md:133]:
```suggestion
Following the reasoning from [Writing a Java library with better experience (slide 12)](https://speakerdeck.com/trustin/writing-a-java-library-with-better-experience?slide=12), 
usage of `java.util.Optional` is kept at a minimum in this project.
```

**Review by theletterf — APPROVED** [review]
Approved with a couple suggestions.

File: docs/contributing/style-guideline.md (hunk 1 above)
Comment (trask) [reply docs/contributing/style-guideline.md:136]: maybe `It is acceptable to use`? we don't exactly want to "recommend" it, only mention that it is "ok" (i.e. author's choice to use or not)

**Pull request merged**